src/halmos/calldata.py

```diff
@@ -210,3 +247,71 @@ def head_size(x):
         static = len(tails) == 0
 
         return EncodingResult(heads + tails, total_size, static)
+
+
+@dataclass(frozen=True)
+class FunctionInfo:
+    name: str | None = None
+    sig: str | None = None
+    selector: str | None = None
+
+
+def str_abi(item: dict) -> str:
+    def str_tuple(args: list) -> str:
+        ret = []
+        for arg in args:
+            typ = arg["type"]
+            match = re.search(r"^tuple((\[[0-9]*\])*)$", typ)
+            if match:
+                ret.append(str_tuple(arg["components"]) + match.group(1))
+            else:
+                ret.append(typ)
+        return "(" + ",".join(ret) + ")"
+
+    if item["type"] != "function":
+        raise ValueError(item)
+    return item["name"] + str_tuple(item["inputs"])
+
+
+def find_abi(abi: list, fun_info: FunctionInfo) -> dict:
+    funname, funsig = fun_info.name, fun_info.sig
+    for item in abi:
+        if (
+            item["type"] == "function"
+            and item["name"] == funname
+            and str_abi(item) == funsig
+        ):
+            return item
+    raise ValueError(f"No {funsig} found in {abi}")
+
+
+def mk_calldata(
+    abi: list,
+    fun_info: FunctionInfo,
+    cd: ByteVec,
+    dyn_param_size: DynamicParams,
+    args: HalmosConfig,
+    arrlen: dict[str, int] = None,
+    new_symbol_id: Callable = None,
+) -> None:
+    # find function abi
+    fun_abi = find_abi(abi, fun_info)
+
+    # no parameters
+    if len(fun_abi["inputs"]) == 0:
+        return
+
+    # generate symbolic ABI calldata
+    arrlen = mk_arrlen(args) if arrlen is None else arrlen
+    calldata = Calldata(args, arrlen, dyn_param_size, new_symbol_id)
+    calldata.create(fun_abi, cd)
+
+
+def mk_arrlen(args: HalmosConfig) -> dict[str, int]:
+    arrlen = {}
+    if args.array_lengths:
+        for assign in [x.split("=") for x in args.array_lengths.split(",")]:
+            name = assign[0].strip()
+            size = assign[1].strip()
+            arrlen[name] = int(size)
+    return arrlen
```

Comment:
dict comprehension would nice here IMO:

```suggestiondef mk_arrlen(args: HalmosConfig) -> dict[str, int]:
    if not args.array_lengths:
        return {}

    name_size_pairs = args.array_lengths.split(",")
    return {
        name.strip(): int(size.strip())
        for name, size in (x.split("=") for x in name_size_pairs)
    }
```

Reply:
accepted you suggestion in https://github.com/a16z/halmos/pull/360/commits/7124320edcaf4534395ddec2dddf2f8c70f1b236, but replaced the tuple-generator comprehension with list comprehension. please let me know if the tuple generator was intentional and performs better than list comprehension here.